### PR TITLE
Fixed a bug that caused compressed images with TFORM missing the optional '1' prefix to not be readable

### PR DESCRIPTION
**Pull request by astrofrog**
Fixes https://github.com/astropy/astropy/issues/14985

### CODE REVIEW

**Comment by astrofrog** [comment]
pre-commit.ci autofix

File: astropy/io/fits/_tiled_compression/tiled_compression.py [hunk 1]
```diff
@@ -203,12 +203,13 @@ def _check_compressed_header(header):
             if not np.isreal(header[kw]) or not float(header[kw]).is_integer():
                 raise TypeError(f"{kw} should be an integer")
 
-    if "TFORM1" in header:
-        for valid in ["1PB", "1PI", "1PJ", "1QB", "1QI", "1QJ"]:
-            if header["TFORM1"].startswith(valid):
-                break
-        else:
-            raise RuntimeError(f"Invalid TFORM1: {header['TFORM1']}")
+    for suffix in range(1, 10):
```
Comment (saimn) [inline astropy/io/fits/_tiled_compression/tiled_compression.py:206]: Not sure if it's possible to have more than 2 columns ? (`COMPRESSED_DATA` and `GZIP COMPRESSED DATA`) ?

File: astropy/io/fits/_tiled_compression/tiled_compression.py [hunk 2]
```diff
@@ -253,11 +254,13 @@ def _get_compression_setting(header, name, default):
...
+    if tform.startswith("1"):
+        tform = tform[1:]
+    if tform[1] == "B":
         dtype = np.uint8
-    elif tform[2] == "I":
+    elif tform[1] == "I":
         dtype = ">i2"
-    elif tform[2] == "J":
+    elif tform[1] == "J":
         dtype = ">i4"
```
Comment (saimn) [inline astropy/io/fits/_tiled_compression/tiled_compression.py:264]: Actually this if/else block can probably be replaced by `tform.recformat.dtype`.

File: astropy/io/fits/_tiled_compression/tiled_compression.py (hunk 1 above)
Comment (astrofrog) [reply astropy/io/fits/_tiled_compression/tiled_compression.py]: One can also have ZSCALE and ZZERO for instance, and I'm not sure if the standard disallows additional columns that the user might have added - but I've now changed this to use TFIELDS to know how many columns there actually are.

File: astropy/io/fits/_tiled_compression/tiled_compression.py (hunk 2 above)
Comment (astrofrog) [reply astropy/io/fits/_tiled_compression/tiled_compression.py:264]: I just tried this but got some test failures, so I might leave this to another follow-up PR as I need time to investigate what is going wrong.

**Comment by astrofrog** [comment]
pre-commit.ci autofix

File: astropy/io/fits/_tiled_compression/tiled_compression.py (hunk 1 above)
Comment (saimn) [reply astropy/io/fits/_tiled_compression/tiled_compression.py]: Seems unlikely that there would be more columns:

> 10.1.3. Table columns
Two columns in the FITS binary table are defined below to con-
tain the compressed image tiles; the order of the columns in the
table is not significant.
[...]
COMPRESSED DATA – [variable-length; required]
GZIP COMPRESSED DATA [variable-length; optional]

But using TFIELDS seems a fine solution :+1:

File: astropy/io/fits/_tiled_compression/tiled_compression.py (hunk 2 above)
Comment (saimn) [reply astropy/io/fits/_tiled_compression/tiled_compression.py:264]: Sure, it's just that it seems we could reuse more code from `io.fits.column` for the header parsing and maybe for the verification, but fine to keep this for later.

**Review by saimn — APPROVED** [review]
Thanks @astrofrog !

**Comment by pllim** [comment]
That's weird. Why would auto backport fail? Is this not applicable to 5.3.x?

**Comment by pllim** [comment]
@astrofrog , you changed the API signature in https://github.com/astropy/astropy/pull/14821 . So this cannot be backported cleanly without also backporting that one. I am not sure if I want to make assumptions here, so I'll leave the manual backport to you.

**Comment by astrofrog** [comment]
Manual backport PR: https://github.com/astropy/astropy/pull/15014

**Pull request merged**